File: app/python/MusicPlayer/view_models/recommendation_song_lists_view_model.py
from dataclasses import dataclass, field
from datetime import datetime
from enum import IntEnum
from typing import Dict, List, Any, Union
import asyncio
from PyQt5.QtCore import QAbstractListModel, QModelIndex, Q_ENUM, pyqtProperty

from MusicPlayer.api.api import API
from MusicPlayer.api.data.personalized import APIPersonalizedItemData, APIPersonalizedData
from MusicPlayer.datetime_util import local_now_with_timezone
from MusicPlayer.view_models.base import BaseViewModel


class _Dependency:
    def __init__(self, api: API):
        self.api = api


@dataclass
class _EveryDayRecommendation:
    name: str = field(init=False, default="每日歌曲推荐")
    tips: str = field(init=False, default="根据您的音乐口味生成每日更新")
    today: datetime = field(init=False, default_factory=local_now_with_timezone)


class _ItemKind(IntEnum):
    Normal = 0
    EveryDay = 1


class _Roles(IntEnum):
    imageUrl = 0
    id = 1
    name = 2
    playCount = 3
    tips = 4
    today = 5
    kind = 6


_ModelItem = Union[_EveryDayRecommendation, APIPersonalizedItemData]


class _RecommendationSongsListModel(QAbstractListModel):
    _list: List[_ModelItem]

    def __init__(self, parent=None):
        super().__init__(parent)
        self._list = []

    def roleNames(self) -> Dict[int, 'QByteArray']:
        roles = dict()

        for name, member in _Roles.__members__.items():
            roles[member.value] = name.encode()

        return roles

    def rowCount(self, parent: QModelIndex) -> int:
        return len(self._list)

    def reset_api_data(self, data: APIPersonalizedData):
        self.beginResetModel()
        self._list.clear()
        # self._list.append(_EveryDayRecommendation())
        self._list.extend(data.result)
        self.endResetModel()

    def data(self, index: QModelIndex, role: int) -> Any:
        item = self._list[index.row()]
        if type(item) == APIPersonalizedItemData:
            if role == _Roles.imageUrl:
                return item.pic_url
            elif role == _Roles.id:
                return item.id
            elif role == _Roles.name:
                return item.name
            elif role == _Roles.playCount:
                if item.play_count > 10000:
                    return f"{int(item.play_count/10000)}万"
                else:
                    return f"{int(item.play_count)}"
            elif role == _Roles.kind:
                return _ItemKind.Normal
            elif role == _Roles.tips:
                return None
            elif role == _Roles.today:
                return None

        if type(item) == _EveryDayRecommendation:
            if role == _Roles.imageUrl:
                return None
            elif role == _Roles.id:
                return -1
            elif role == _Roles.name:
                return item.name
            elif role == _Roles.playCount:
                return None
            elif role == _Roles.kind:
                return _ItemKind.EveryDay
            elif role == _Roles.tips:
                return item.tips
            elif role == _Roles.today:
                return item.today.day


class RecommendationSongListsViewModel(BaseViewModel):
    _d: _Dependency

    Q_ENUM(_ItemKind)

    def __init__(self, parent=None):
        super().__init__(_Dependency, parent)
        self._list_model = _RecommendationSongsListModel(self)

    def componentComplete(self):
        asyncio.create_task(self.load_recommendation_list())
        super().componentComplete()

    async def load_recommendation_list(self):
        data = await self._d.api.personalized(10)
        self._list_model.reset_api_data(data)

    @pyqtProperty(QAbstractListModel)
    def songLists(self):
        return self._list_model

File: app/cpp/MusicPlayer/view_models/recommendation_song_lists_view_model.cc
//
// Created by rick on 2019/11/10.
//

#include "recommendation_song_lists_view_model.h"
#include "../api/api.h"
#include "../asyncfuture.h"
#include <QtCore/QAbstractListModel>

namespace MusicPlayer::ViewModels {
    using namespace MusicPlayer::API;

    static QString kWeekDays[] = {
        QStringLiteral(u"星期一"), QStringLiteral(u"星期二"),
        QStringLiteral(u"星期三"), QStringLiteral(u"星期四"),
        QStringLiteral(u"星期五"), QStringLiteral(u"星期六"),
        QStringLiteral(u"星期日"),
    };

    class RecommendationSongsListModel : public QAbstractListModel {
        Q_OBJECT
      public:
        using ModeItem =
            std::variant<EveryDayRecommendation, APIPersonalizedItemData>;
        enum Roles {
            ImageUrl = Qt::UserRole + 1,
            Id,
            Name,
            PlayCount,
            Tips,
            Today,
            Weekday,
            Kind
        };

        explicit RecommendationSongsListModel(QObject* parent = nullptr);

        int rowCount(const QModelIndex& index) const override;

        QHash<int, QByteArray> roleNames() const override;

        QVariant data(const QModelIndex& index, int role) const override;

        void setRecommendationSongListsData(const APIPersonalizedData& data) {
            beginResetModel();
            _songsList.clear();
            _songsList.append(EveryDayRecommendation{
                data.result.last().value<APIPersonalizedItemData>().picUrl});
            std::transform(data.result.cbegin(), data.result.cend() - 1,
                           std::back_inserter(_songsList), [](const auto& var) {
                               return var.template value<APIPersonalizedItemData>();
                           });
            endResetModel();
        }

      private:
        QVector<ModeItem> _songsList;
    };

    QVariant RecommendationSongsListModel::data(const QModelIndex& index,
                                                int role) const {
        const auto& item = _songsList[index.row()];

        return std::visit(
            [=](const auto& value) -> QVariant {
                if constexpr (std::is_convertible_v<decltype(value),
                                                    APIPersonalizedItemData>) {
                    if (role == Roles::ImageUrl)
                        return value.picUrl;
                    if (role == Roles::Id)
                        return value.id;
                    if (role == Roles::Name)
                        return value.name;
                    if (role == Roles::PlayCount) {
                        if (value.playCount > 10000)
                            return QString(QStringLiteral(u"%1万"))
                                .arg(value.playCount / 10000);
                        else
                            return QString("%1").arg(value.playCount);
                    }
                    if (role == Roles::Kind)
                        return RecommendationSongListsViewModel::Normal;

                    return QVariant();
                } else {
                    if (role == Roles::Id)
                        return -1;
                    if (role == Roles::Name)
                        return value.name;
                    if (role == Roles::Kind)
                        return RecommendationSongListsViewModel::EveryDay;
                    if (role == Roles::Tips)
                        return value.tips;
                    if (role == Roles::Today)
                        return value.today.date().day();
                    if (role == Roles::Weekday)
                        return kWeekDays[value.today.date().dayOfWeek() - 1];
                    if (role == Roles::ImageUrl)
                        return value.imageUrl;
                    return QVariant();
                }
            },
            item);
    }

    QHash<int, QByteArray> RecommendationSongsListModel::roleNames() const {
        QHash<int, QByteArray> names;

        names[Roles::ImageUrl]  = "imageUrl";
        names[Roles::Id]        = "id";
        names[Roles::Name]      = "name";
        names[Roles::PlayCount] = "playCount";
        names[Roles::Tips]      = "tips";
        names[Roles::Today]     = "today";
        names[Roles::Weekday]   = "weekday";
        names[Roles::Kind]      = "kind";

        return names;
    }

    int RecommendationSongsListModel::rowCount(const QModelIndex& index) const {
        int c = _songsList.size();
        return c;
    }

    RecommendationSongsListModel::RecommendationSongsListModel(QObject* parent)
        : QAbstractListModel{parent} {}

    RecommendationSongListsViewModel::RecommendationSongListsViewModel(
        QObject* parent)
        : QObject{parent} {
        _songsModel = new RecommendationSongsListModel(this);
    }

    void RecommendationSongListsViewModel::loadData() {
        MusicAPI api;
        auto f = api.personalized(10);
        AsyncFuture::observe(f).subscribe([this](const Response<
                                                 APIPersonalizedData>& reply) {
            std::visit(
                [=](const auto& value) {
                    if constexpr (std::is_convertible_v<decltype(value),
                                                        APIPersonalizedData>) {
                        _songsModel->setRecommendationSongListsData(value);
                    } else {
                        // errors
                    }
                },
                reply);
        });
    }

    void RecommendationSongListsViewModel::componentComplete() { loadData(); }

    QAbstractListModel* RecommendationSongListsViewModel::songLists() {
        return _songsModel;
    }
} // namespace MusicPlayer::ViewModels

#include "recommendation_song_lists_view_model.moc"
